File: lib/ts/src/Serializer.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { Message } from './Message.js';
import { Serialization } from './Serialization.js';
import { BinaryEncoder } from './internal/binary/BinaryEncoder.js';
import { serializeInternal } from './internal/SerializeInternal.js';
import { deserializeInternal } from './internal/DeserializeInternal.js';
import { Base64Encoder } from './internal/binary/Base64Encoder.js';

export class Serializer {
    /**
     * A serializer that converts a Message to and from a serialized form.
     */

    private serializationImpl: Serialization;
    private binaryEncoder: BinaryEncoder;
    private base64Encoder: Base64Encoder;

    constructor(serializationImpl: Serialization, binaryEncoder: BinaryEncoder, base64Encoder: Base64Encoder) {
        this.serializationImpl = serializationImpl;
        this.binaryEncoder = binaryEncoder;
        this.base64Encoder = base64Encoder;
    }

    public serialize(message: Message): Uint8Array {
        /**
         * Serialize a Message into a byte array.
         */
        return serializeInternal(message, this.binaryEncoder, this.base64Encoder, this.serializationImpl);
    }

    public deserialize(messageBytes: Uint8Array): Message {
        /**
         * Deserialize a Message from a byte array.
         */
        return deserializeInternal(messageBytes, this.serializationImpl, this.binaryEncoder, this.base64Encoder);
    }
}

File: lib/go/Serializer.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package telepact

import (
	telepactinternal "github.com/telepact/telepact/lib/go/internal"
	"github.com/telepact/telepact/lib/go/internal/binary"
)

// Serializer converts Telepact messages between pseudo-JSON and serialized bytes.
type Serializer struct {
	serializationImpl Serialization
	binaryEncoder     binary.BinaryEncoder
	base64Encoder     binary.Base64Encoder
}

// NewSerializer constructs a Serializer with the provided serialization and encoding strategies.
func NewSerializer(serializationImpl Serialization, binaryEncoder binary.BinaryEncoder, base64Encoder binary.Base64Encoder) *Serializer {
	return &Serializer{
		serializationImpl: serializationImpl,
		binaryEncoder:     binaryEncoder,
		base64Encoder:     base64Encoder,
	}
}

// Serialize converts a Message into its serialized byte representation.
func (s *Serializer) Serialize(message Message) ([]byte, error) {
	return telepactinternal.SerializeInternal(
		message.Headers,
		message.Body,
		s.binaryEncoder,
		s.base64Encoder,
		s.serializationImpl,
		func(err error, context string) error {
			return NewSerializationError(err, context)
		},
	)
}

// Deserialize converts serialized bytes into a Message structure.
func (s *Serializer) Deserialize(messageBytes []byte) (Message, error) {
	headers, body, err := telepactinternal.DeserializeInternal(messageBytes, s.serializationImpl, s.binaryEncoder, s.base64Encoder)
	if err != nil {
		return Message{}, err
	}
	return NewMessage(headers, body), nil
}
